### PR TITLE
Rename file should rename matching all symbols in the file.

### LINKED ISSUE #272: Rename File should rename all instance of matching symbols in the file
Currently if the file has code like below 

```
namespace namespace1
{
    class foo
    {
        static void Main(string[] args)
        {
        }
    }
}

namespace namespace2
{
    class foo
    {
        static void test()
        {
        }
    }
}
```

Actual: Rename of the file trigger rename of only first symbol
Expected: Rename of the file should rename all matching symbols in the file

### PR DESCRIPTION
**Pull request by jinujoseph**
fixes #272

### CODE REVIEW

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs [hunk 1]
```diff
@@ -75,55 +77,81 @@ internal class Renamer
...
+        private async Task<Solution> GetRenamedSolutionAsync(Project myNewProject)
+        {
+            var project = myNewProject;
+            Solution renamedSolution = null;
+            string oldName = Path.GetFileNameWithoutExtension(_oldFilePath);
 
-            var symbol = semanticModel.GetDeclaredSymbol(declaration);
-            if (symbol == null)
-                return;
+            while (true)
```
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs]: Should this be while (project != null)? Otherwise since you're calling FirstOrDefault in the last line of this loop block, the next time around the first line of loop block would blow up.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/RenamerTests.cs [hunk 2]
```diff
@@ -20,10 +22,38 @@ public class RenamerTests
         [InlineData("partial class Foo {} partial class Foo {}", "Foo.cs", "Bar.cs")]
         [InlineData("struct Foo { decimal price; string title; string author;}", "Foo.cs", "Bar.cs" )]
         [InlineData("enum Foo { None, enum1, enum2, enum3, enum4 };", "Foo.cs", "Bar.cs")]
+        [InlineData("namespace n1 {class Foo{}} namespace n2 {class Foo{}}", "Foo.cs", "Bar.cs")]
+        public async Task Rename_Symbol_Should_TriggerUserConfirmationAsync(string soureCode, string oldFilePath, string newFilePath)
+        {
+            var userNotificationServices = IUserNotificationServicesFactory.Implement(f =>
```
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/RenamerTests.cs]: The Moq framework records all the calls that are made on the mocked object. So you can simply do Mock.Get(userNotificationServices).Verify(f => f.Rename()) to get the Mock object from the instance and verify that the Rename method was called.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/Mocks/IRoslynServicesFactory.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS
+{
+    internal static class IRoslynServicesFactory
+    {
+        public static IRoslynServices Implement(Func<Solution, ISymbol, string, Task<Solution>> renameSymbolAsync, Func<Workspace,Solution,bool> applyChangesToSolution)
+        {
+            var mock = new Mock<IRoslynServices>();
+           
+            mock.Setup(h => h.RenameSymbolAsync(It.IsAny<Solution>(), It.IsAny<ISymbol>(), It.IsAny<string>()))
```
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/Mocks/IRoslynServicesFactory.cs]: Looks like you need these Funcs just to record the fact that it was called. If you use the Verify method on the mock object, you can get simplify all of this.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs [hunk 4]
```diff
@@ -75,55 +77,81 @@ internal class Renamer
...
+
+                var semanticModel = await newDocument.GetSemanticModelAsync().ConfigureAwait(false);
+                if (semanticModel == null)
+                    return renamedSolution;
+
+                var symbol = semanticModel.GetDeclaredSymbol(declaration);
+                if (symbol == null)
+                    return renamedSolution;
+
+                bool userConfirmed = await CheckUserConfirmation();
```
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs]: So we'll pop up a dialog for every declaration in the file? That doesn't sound good.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs [hunk 5]
```diff
@@ -75,55 +77,81 @@ internal class Renamer
...
+                Document newDocument = (from d in project.Documents where StringComparers.Paths.Equals(d.FilePath, _newFilePath) select d).FirstOrDefault();
+                if (newDocument == null)
+                    return renamedSolution;
+                
+                var root = await newDocument.GetSyntaxRootAsync().ConfigureAwait(false);
+                if (root == null)
+                    return renamedSolution;
+
+                var declarations = root.DescendantNodes().Where(n => HasMatchingSyntaxNode(newDocument, n, oldName));
+                var declaration = declarations.FirstOrDefault();
```
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs]: Why not loop over these declarations instead of putting everything inside a while loop?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/RenamerTests.cs (hunk 2 above)
Comment (jinujoseph) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/RenamerTests.cs]: oh, good to know ...will use that

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs (hunk 4 above)
Comment (jinujoseph) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs]: No we won't _userPromptedOnce takes care of that

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs (hunk 5 above)
Comment (jinujoseph) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs]: Tried that, did not work ..since RenameSymbolAsync returns me a new snapshot ..i lose the reference to the declaration.

Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs]: Ah yes.. You can add an annotation to the nodes and then call tree.GetAnnotatedNodes(myAnnotation) to get them back. However, that's probably more expensive than what you're doing. Can you please add some comments explaining this since it's not obvious from looking at the code.

Comment (jinujoseph) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs]: sure , will do

**Comment by jinujoseph** [comment]
@srivatsn  Addressed the feedback

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/Mocks/IRoslynServicesFactory.cs [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+using System;
+using Moq;
+using Microsoft.CodeAnalysis;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS
+{
+    internal static class IRoslynServicesFactory
+    {
+        public static Mock<IRoslynServices> Create()
```
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/Mocks/IRoslynServicesFactory.cs]: In general the pattern is that we have the factories return the actual service itself (i.e. mock.Object) and then you can use Mock.Get(object) to get back the mock (for implement or Verify). It's odd to have to know that the Factory for x doesn't generate x.

Comment (jinujoseph) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/Mocks/IRoslynServicesFactory.cs]: fixed

**Comment by srivatsn** [comment]
👍

**Comment by jinujoseph** [comment]
@dotnet-bot retest this please

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/RenamerTests.cs [hunk 7]
```diff
@@ -5,27 +5,54 @@
...
     {
-        bool ConfirmForRenameWasCalled { get; set; }
+        [Theory]
+        [InlineData("class Foo{}", "Foo.cs", "Bar.cs")]
+        [InlineData("interface Foo { void m1(); void m2();}", "Foo.cs", "Bar.cs")]
+        [InlineData("delegate int Foo(string s);", "Foo.cs", "Bar.cs")]
+        [InlineData("partial class Foo {} partial class Foo {}", "Foo.cs", "Bar.cs")]
+        [InlineData("struct Foo { decimal price; string title; string author;}", "Foo.cs", "Bar.cs")]
+        [InlineData("enum Foo { None, enum1, enum2, enum3, enum4 };", "Foo.cs", "Bar.cs")]
+        [InlineData("namespace n1 {class Foo{}} namespace n2 {class Foo{}}", "Foo.cs", "Bar.cs")]
```
Comment (basoundr) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/RenamerTests.cs]: The test cases in a single line is easy to understand for cases where there is a single level of nesting. For Eg: The first testcase, `class Foo{}`, is easy to understand. But it gets difficult to understand what the code looks like here `namespace n1 {class Foo{}} namespace n2 {class Foo{}}`. Not very difficult but if the code is written in the same way as we would write with line breaks, indents and everything, it will be very easy to understand. I understand that we haven't reached that high level of complexity for test code but this is something to keep in mind as we implement new feature and write more tests

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/RenamerTests.cs [hunk 8]
```diff
@@ -5,27 +5,54 @@
...
+
+            Mock.Get(userNotificationServices).Verify(h => h.Confirm(It.IsAny<string>()), Times.Once);
+            Mock.Get(roslynServices).Verify(h => h.RenameSymbolAsync(It.IsAny<Solution>(), It.IsAny<ISymbol>(), It.IsAny<string>()), Times.Never);
+        }
 
         [Theory]
         [InlineData("class Foo{}", "Foo.cs", "Bar.cs")]
         [InlineData("interface Foo { void m1(); void m2();}", "Foo.cs", "Bar.cs")]
         [InlineData("delegate int Foo(string s);", "Foo.cs", "Bar.cs")]
         [InlineData("partial class Foo {} partial class Foo {}", "Foo.cs", "Bar.cs")]
```
Comment (basoundr) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/RenamerTests.cs]: As I am looking into the tests, I see that there is no assertion for the end result. Am I missing to see it? There is no assertion to check the text change. For the most part it might be testing Roslyn Rename but it helps us to make sure we renamed all the required symbols and also not rename the ones we shouldn't.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs [hunk 9]
```diff
@@ -75,55 +78,83 @@ internal class Renamer
 
         public async Task RenameAsync(Project myNewProject)
         {
-            Document newDocument = (from d in myNewProject.Documents where StringComparers.Paths.Equals(d.FilePath, _newFilePath) select d).FirstOrDefault();
-            if (newDocument == null)
+            Solution renamedSolution = await GetRenamedSolutionAsync(myNewProject).ConfigureAwait(false);
+            if (renamedSolution == null)
                 return;
```
Comment (basoundr) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs]: remove extra space

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs [hunk 10]
```diff
@@ -75,55 +78,83 @@ internal class Renamer
...
+            {
+                Document newDocument = (from d in project.Documents where StringComparers.Paths.Equals(d.FilePath, _newFilePath) select d).FirstOrDefault();
+                if (newDocument == null)
+                    return renamedSolution;
+                
+                var root = await newDocument.GetSyntaxRootAsync().ConfigureAwait(false);
+                if (root == null)
+                    return renamedSolution;
+
+                var declarations = root.DescendantNodes().Where(n => HasMatchingSyntaxNode(newDocument, n, oldName));
```
Comment (basoundr) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs]: inline

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs (hunk 9 above)
Comment (jinujoseph) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs]: I think this is Git messing some thing up, i don't see an extra line

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/RenamerTests.cs (hunk 8 above)
Comment (jinujoseph) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/RenamerTests.cs]: Sure , Makes sense

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/RenamerTests.cs (hunk 7 above)
Comment (jinujoseph) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/RenamerTests.cs]: Sure , Makes sense

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs (hunk 9 above)
Comment (basoundr) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs]: I pulled your change to my machine and I was able to see the spaces

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/RenamerTests.cs (hunk 8 above)
Comment (basoundr) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/RenamerTests.cs]: @jinu roslynService ensures the correctness of the rename operation. But our code initiates the rename so we must make sure we dont rename the inappropriate symbol and rename only the required ones.

Comment (jinujoseph) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/RenamerTests.cs]: @basoundr , as discussed will send a separate PR with it.
